Enable -Wstring-concatenation and -Wstring-conversion on clang builds

Empact opened the pull request:
The only current violations are a few uses of assert(!"descriptive string").

Seems the benefit of ensuring against a code fault outweighs the benefit of possible additional information in these cases.

> [-Wstring-concatenation](https://clang.llvm.org/docs/DiagnosticsReference.html#wstring-concatenation)
> Diagnostic text:
>
> warning: suspicious concatenation of string literals in an array initialization; did you mean to separate the elements with a comma?
>
> [-Wstring-conversion](https://clang.llvm.org/docs/DiagnosticsReference.html#id805)
> Diagnostic text:
>
> warning: implicit conversion turns string literal into bool: A to B

https://clang.llvm.org/docs/DiagnosticsReference.html#wstring-concatenation

Empact committed:
Enable -Wstring-concatenation and -Wstring-conversion on clang builds

The only current violations are a few uses of assert(!"descriptive string").

Seems the benefit of ensuring against a code fault outweighs the benefit of
possible additional information in these cases.

theuni commented on src/net_processing.cpp:
```diff
@@ -2141,17 +2141,15 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
-            assert(!"cannot load block from disk");
-        }
+        // fail if cannot load block from disk
+        assert(ReadRawBlockFromDisk(block_data, pindex->GetBlockPos(), m_chainparams.MessageStart()));
```
Could you please split this up like this instead?
```c++
auto foo = ReadRawBlockFromDisk(...);
assert(foo);
```
That way the we don't introduce a new side-effect of `NDEBUG=1`

theuni commented on src/net_processing.cpp:
```diff
@@ -2141,17 +2141,15 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
-            assert(!"cannot load block from disk");
-        }
+        // fail if cannot load block from disk
+        assert(ReadBlockFromDisk(*pblockRead, pindex, m_chainparams.GetConsensus()));
```
Likewise here.

Empact replied:
Done

theuni commented on src/script/interpreter.cpp:
```diff
@@ -1464,14 +1464,11 @@ const HashWriter HASHER_TAPBRANCH{TaggedHash("TapBranch")};
 
 static bool HandleMissingData(MissingDataBehavior mdb)
 {
-    switch (mdb) {
```
Imo it'd be cleaner and more readable to just leave the logic here as before, turning the wonky string assertions into comments.

Empact replied:
Done

theuni reviewed (changes requested):
Concept ACK. This has been open for a while, I'm curious to know if any new violations have snuck in since?

theuni commented:
Thanks! Now mind updating to use `Assert()`? :)

Empact committed:
refactor: Convert string-conversion violation asserts to use Assert from util/check

Empact commented:
Rebased, converted assignments to initialization, and moved to `Assert` where possible. I could not apply `Assert(false)` in interpreter.cpp due to errors, but https://github.com/bitcoin/bitcoin/pull/26504 could be applied here once merged.
